### DISCUSSION TITLE
Cara menambahkan kolom baru (dengan kriteria tertentu)

### DISCUSSION THREAD

**Discussion by yohanespm77** [teamalgoritma/community#541] [Q&A] [answered]
Saya punya data seperti ini :
<img width="327" alt="image" src="https://user-images.githubusercontent.com/106010284/218372018-b6c28967-b89b-489e-b60c-a67ff80b4d54.png">

Bagaimana cara membuat kolom baru yang isinya keterangan dari variabel "trx_time" dengan ketentuan : 
- apabila di hari yg sama dan nama yg sama ada jam yg berada di antara 00:00:01 s/d 05:00:00 maka didefine sebagai "1"
- apabila ada data berikutnya di hari yg sama dan nama yg sama di jam antara 00:00:01 s/d 05:00:00 maka didefine sebagai "2"
- diluar jam 00:00:01 s/d 05:00:00 didefine sebagai "0"

Outputnya mau seperti ini : 
<img width="465" alt="image" src="https://user-images.githubusercontent.com/106010284/218374290-5c5afb62-8cf8-457b-988b-a2013fa3442c.png">

Saya sudah coba buat seperti ini, tapi masih belum sesuai outputnya 
<img width="547" alt="image" src="https://user-images.githubusercontent.com/106010284/218374183-525d13f5-a459-4fd3-a95c-24ce7a20ad1d.png">

Mohon bantuannya kak, terima kasih.

**Comment by finesaaa** [comment, marked as answer]
Dear Pak @yohanespm77 

Untuk melakukan hal seperti itu yang memiliki berbagai kondisi grouping, kita memerlukan beberapa steps, tidak cukup dengan ifelse yang hanya bisa menangani 2 kondisi

1. Bapak perlu membuat sebuah dataframe untuk melakukan countur terhadap setiap kostumer pada setiap hari yang ada dengan cara berikut
```
library(dplyr)
options(dplyr.summarise.inform = FALSE)
counter <- data %>% group_by(date=as_date(trx_time), name) %>% summarise()
counter['count'] <- rep(0, nrow(counter))
```
![image](https://user-images.githubusercontent.com/57583780/218431884-a2138e00-1d51-486b-8fde-3c27466e5294.png)

2. Selanjutnya counter ini akan membantu bapak untuk melakukan counter pada setiap orang di hari yang sama. Saya akan bantu membuat perulangan untuk melakukan request yang Bapak minta
```
library(lubridate)
jam00_jam05 <- c()

for (row in 1:nrow(data)) {
  tmp = data[row,]
  if (hour(tmp$trx_time) >= 00 & hour(tmp$trx_time) <= 05) {
    cond <- (tmp$name == counter$name & as_date(tmp$trx_time) == counter$date)
    counter[cond, ]$count <- counter[cond, ]$count + 1
    
    jam00_jam05 <- append(jam00_jam05, counter[cond, ]$count)
  } else {
    jam00_jam05<- append(jam00_jam05, 0)
  }
}

data['jam00_jam05'] <- jam00_jam05
```
![image](https://user-images.githubusercontent.com/57583780/218432243-981e2668-197a-4fb9-8064-9f024fe483ec.png)

Sehingga akan menghasilkan data seperti gambar di atas pak. 
Semoga membantu, terima kasih 🙏

Sebaiknya kedua step bapak lakukan secara bersamaan dalam satu cell, agar counternya selalu tereset 0 sebelum bapak melakukan grouping

**Reply by yohanespm77** [reply]
Terima kasih kak atas jawabannya. 

Tapi utk yg jam 00:00:00 kedefine sebagai 1 kak, padahal harusnya 0.
<img width="568" alt="image" src="https://user-images.githubusercontent.com/106010284/218521899-3ddb5192-b05a-4513-9671-11b3debf4b46.png">

Kalau ga pakai "=" jadi kedefine sebagai 0
<img width="527" alt="image" src="https://user-images.githubusercontent.com/106010284/218522549-69c3e401-683e-438e-9001-cc8b39295af9.png">

Pertanyaan saya : 
1) Agar yg jam 00:00:00 tetap kedefine sebagai 0 gimana ya kak ?
2) if (hour(tmp$trx_time) >= 00 & hour(tmp$trx_time) <= 05) 
Itu kan fungsinya pake "hour" ya kak. Kalo mau lebih spesifik lagi, misal antara jam 00:30:25 s/d 04:45:33 gimana ya kak ?

**Reply by finesaaa** [reply]
Dear pak @yohanespm77,
Terkait hal itu bapak bisa coba memberikan lebih spesifik untuk conditionalnya dengan mengubah baris
```
if (hour(tmp$trx_time) >= 00 & hour(tmp$trx_time) <= 05)
```
Menjadi
```
if (hms::as.hms(tmp$trx_time) >= hms::as.hms('00:30:25') & hms::as.hms(tmp$trx_time) <= hms::as.hms('04:45:33')) 
```
Sehingga dapat menghasilkan hasil sebagai berikut
![image](https://user-images.githubusercontent.com/57583780/218635514-6888b15c-042f-41f0-aa22-069346a77c7b.png)

Sekian, semoga membantu pak.

**Reply by yohanespm77** [reply]
Baik kak, terima kasih banyak kak @finesaaa . Jawabannya sangat membantu 👍
